src/sparknacl-mac.adb

```diff
@@ -191,4 +191,35 @@ is
       return Equal (H, X);
    end Onetimeauth_Verify;
 
+   --------------------------------------------------------
+   --  HMAC
+   --------------------------------------------------------
+
+   procedure HMAC_SHA_256 (Output :    out Hashing.Digest_256;
+                           M      : in     Byte_Seq;
+                           K      : in     Byte_Seq)
+   is
+      IPad : Bytes_64 := (others => 16#36#);
+      OPad : Bytes_64 := (others => 16#5C#);
+
+      --  K', pad to the right with zeroes if shorter than Block_Size, or
+      --  Hash down to less than block size first if greater than Block_Size.
+      Key : Bytes_64 := (others => 0);
+   begin
+      if K'Length > 64 then
+         Key (0 .. 31) := Hashing.Hash_256 (K);
+      else
+         pragma Assert (K'Length <= 64);
+         Key (K'Range) := K;
+      end if;
+
+      for I in Index_64'Range loop
+         IPad (I) := IPad (I) xor Key (I);
+         OPad (I) := OPad (I) xor Key (I);
+      end loop;
+
+      Output := Hashing.Hash_256 (OPad & Hashing.Hash_256 (IPad & M));
```

Comment:
Yeah the concatenations are troublesome. Other libraries use an "update" style API which allows you to "build" the hash up with subsequent data. It seems like DJB (designer of ChaCha20/Salsa20 and author of NaCl and TweetNaCl) isn't a big fan of update libraries (https://groups.google.com/g/boring-crypto/c/BpUmNMXKMYQ for a little background), and I can understand the rationale for preferring a "one-shot" library function. 

An alternative might be to allocate a bigger buffer and then copy the results into it at the appropriate offsets, but you'd still need the stack space for the full buffer, so I'm not sure that it would save any stack space. It would have the downside too of being a bit more complicated to prove correctness.

Reply:
I had wondered why NaCl didn't have an "update" style API for hashing... now I know!